### ISSUE TITLE
MismatchingDocblockReturnType not working correctly

### ISSUE THREAD

**Issue by rmikalkenas** [vimeo/psalm#9699]
Hello,
correct me if wrong, but shouldn't such code result to MismatchingDocblockReturnType error?

https://psalm.dev/r/0ca08f8c19

**Comment by orklah** [comment]
You're always allowed to be more precise in the docblock than in the signature type but you're not allowed to contradict it. It's normal that MismachingDocblocReturnType is not raised.

However, I would expect those to be raised: https://psalm.dev/r/a9260e0cbe
I'm not classifying this as a bug yet because there could be an explanation, but this is still weird

**Comment by orklah** [comment]
I'll let this open because [this](https://psalm.dev/r/a9260e0cbe) should be investigated at some point

**Comment by robchett** [comment]
I've tracked this down to this code, where null is being added to the docblock defined return type because it's in the signature but not in the docblock and they are otherwise identical.
https://github.com/vimeo/psalm/blob/2039667e35b953fa43b96682510fc8f279913b15/src/Psalm/Internal/PhpVisitor/Reflector/FunctionLikeDocblockScanner.php#L1019-L1037 

If the docblock intent is supposed to overrule language features then should this be removed? Using a docblock like this does seem like a bad idea as it will mask errors at the call site but also useDocblockTypes exists if you don't want to trust them...

**Comment by robchett** [comment]
Current behaviour appears to be to only throw this Issue if the docblock type is not contained in the signature type and the signature type isn't coerced from mixed.

In this case `Type[]` is contained in `?array` and also `array` in the signature is triggering the coerced from mixed check as it is `array<array-key, mixed>`

I've attempted to try and resolve this but there are to many edge cases for when the docblock type can represent a type the signature type can't due to language limitations.

My current thinking is the issue should be thrown if the different between the types contains only:

- Mixed in the signature
- Array<Mixed> in the signature
- A template in the docblock

If that logic is sound I'll take another stab at this

**Comment by orklah** [comment]
> I've attempted to try and resolve this but there are to many edge cases for when the docblock type can represent a type the signature type can't due to language limitations.

Yeah but being more narrow in the docblock than in the signature is fine. What's not fine is having a wider type returned vs what's described in the docblock and I can't think of a legit case for that. Do you have an edge case in mind?

**Comment by kkmuffme** [comment]
https://github.com/vimeo/psalm/issues/9699#issuecomment-1520678461 seems to be already fixed for both null and any union types (InvalidReturnType) @orklah
